Websocket source

krishnaprasadas opened the pull request:
New Websocket source app.

Resolves spring-cloud-stream-app-starters/websocket#2

krishnaprasadas committed:
test path changed

Commit:
Test and Doc

pivotal-issuemaster commented:
@krishnaprasadas Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud-stream-app-starters/websocket&pullRequestId=3)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud-stream-app-starters/websocket&pullRequestId=3) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@krishnaprasadas Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud-stream-app-starters/websocket&pullRequestId=3)!

krishnaprasadas commented:
The following is missing in the code,
1. I can't sucessfully create test cases there were some dependency issues. So I commented the code.
2. My README.adoc is not generating, on compile, there is nullpointer exception, I think its not a syntax issue because I tried with the same from sink project. So I have renamed the README file.

Commit:
readme file renamed

krishnaprasadas committed:
configuration-properties added

configuration-properties added

krishnaprasadas committed:
Source readme link added

krishnaprasadas commented:
I have updated the README doc which is failing in my dev environment on build (Null Pointer exception). Hope it would work here. Please have a look and comment for any change required.

sabbyanandan commented:
@viniciusccarvalho: When you get a chance, please weigh-in on this proposal.

viniciusccarvalho commented on spring-cloud-starter-stream-source-websocket/src/main/java/org/springframework/cloud/stream/app/websocket/source/WebsocketSourceConfiguration.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * A source module that receives data over WebSocket
+ *
+ * @author krishnaprasad
```
Please change for your full name instead of username on author tags

viniciusccarvalho commented on spring-cloud-starter-stream-source-websocket/src/main/java/org/springframework/cloud/stream/app/websocket/source/WebsocketSourceConfiguration.java:
```diff
@@ -0,0 +1,40 @@
+
+	@Bean
+	public IntegrationWebSocketContainer serverWebSocketContainer() {
+		return new ServerWebSocketContainer(properties.getPath()).setAllowedOrigins("*");
```
We should add `Allowed Origins` as a configurable property

krishnaprasadas replied:
Thank you @viniciusccarvalho for the  comments. Definitely I will fix the issues with your guidance.
About the Readme I'm getting the following error,
```
[ERROR] Failed to execute goal org.springframework.cloud:spring-cloud-app-starter-doc-maven-plugin:1.1.4.RC1:generate-documentation (generate-documentation) on project spring-cloud-starter-stream-source-websocket: Error generating documentation: NullPointerException -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.springframework.cloud:spring-cloud-app-starter-doc-maven-plugin:1.1.4.RC1:generate-documentation (generate-documentation) on project spring-cloud-starter-stream-source-websocket: Error generating documentation
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:216)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:862)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:286)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:197)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.maven.wrapper.BootstrapMainStarter.start(BootstrapMainStarter.java:39)
        at org.apache.maven.wrapper.WrapperExecutor.execute(WrapperExecutor.java:122)
        at org.apache.maven.wrapper.MavenWrapperMain.main(MavenWrapperMain.java:50)
Caused by: org.apache.maven.plugin.MojoExecutionException: Error generating documentation
        at org.springframework.cloud.stream.app.documentation.plugin.ConfigurationMetadataDocumentationMojo.execute(ConfigurationMetadataDocumentationMojo.java:138)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        ... 27 more
Caused by: java.lang.NullPointerException
        at org.springframework.cloud.stream.app.documentation.plugin.ConfigurationMetadataDocumentationMojo.maybeHints(ConfigurationMetadataDocumentationMojo.java:175)
        at org.springframework.cloud.stream.app.documentation.plugin.ConfigurationMetadataDocumentationMojo.asciidocFor(ConfigurationMetadataDocumentationMojo.java:159)
        at org.springframework.cloud.stream.app.documentation.plugin.ConfigurationMetadataDocumentationMojo.execute(ConfigurationMetadataDocumentationMojo.java:119)
        ... 29 more
```
Can you help me to find out the mistake I have done?

viniciusccarvalho reviewed (commented):
@krishnaprasadas I left a few comments on the PR, but there are a couple of very serious issues we need to take care first.

@artembilan mentioned on the issue the fact that you should expose more configurations from the SI support for websocket, and right now `path` is the only one. I requested at least `origins` but users should be able to configure as much as they can given the support from integration, so please change that.

Second point: What's wrong with readme? Just don't rename it and push, that's not a good strategy, you are just moving the problem from your end to ours. If you need help, post the stack trace and we will guide you

Third point: Tests, just like above, just commenting them out, you are just pushing the problem to us, try to fix, post the exception and we will guide you. We can't accept code without testing, and testing is hard but someone has to do it right?

Thank you for the collaboration, it's tough, but in the end rewarding.

artembilan reviewed (changes requested):
Please, reconsider to rebase onto the `master`.

This project now Spring Boot / SCSt 2.0 compatible, therefore your PR should be against actual state.

Note: `spring-integration-java-dsl` doesn't exist any more. Although I don't see why do you need it at all...

Also consider to add Copyright to all your classes.

When you update everything, we will come back to you with more review.

Thank you!

krishnaprasadas commented:
Following changes incorporated,
1. Author name
2. Copyright
3. Conflict resolution with master branch
4. Adding configuration property for 'allowedOrigins'

Please have a look and correct me for any further change if required. I have removed the unit test, will added it once that document generation error got resolved.
Waiting for a help in document generation exception resolution.

Thanks you.

artembilan commented:
> Merge pull request #1 from spring-cloud-stream-app-starters/master 

I didn't talk about `merge`. I asked `rebase`. With the merge you make your own code as mainstream. But when you rebase you don't have choice unless follow the code from the upstream. Now I don't know which code is yours in this PR and which you have just overwritten because of `merge`.

Thank you for understanding.

That indeed might not be so bad. I'll pull your PR locally today and will try to figure out what is the problem you have. However it really would be better to have tests here in PR...

artembilan commented on spring-cloud-starter-stream-source-websocket/README.adoc:
```diff
@@ -0,0 +1,55 @@
+N/A
+
+== Options
+The following commmand line arguments are supported:
```
Typo: triple `m`

artembilan commented on spring-cloud-starter-stream-source-websocket/pom.xml:
```diff
@@ -0,0 +1,69 @@
+	<dependencies>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter</artifactId>
```
I think we don't need this one. It comes transitively from all the Spring Boot starters, e.g. our `spring-boot-starter-websocket` below.

artembilan commented on spring-cloud-starter-stream-source-websocket/pom.xml:
```diff
@@ -0,0 +1,69 @@
+			<dependency>
+				<groupId>org.springframework.cloud.stream.app</groupId>
+				<artifactId>websocket-app-dependencies</artifactId>
+				<version>1.3.0.BUILD-SNAPSHOT</version>
```
Missed to upgrade to `2.0.0`

artembilan commented on spring-cloud-starter-stream-source-websocket/src/main/java/org/springframework/cloud/stream/app/websocket/source/WebsocketSourceConfiguration.java:
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2014-2016 the original author or authors.
```
This is new class and this is new year. So, just `2018`

artembilan commented on spring-cloud-starter-stream-source-websocket/src/main/java/org/springframework/cloud/stream/app/websocket/source/WebsocketSourceConfiguration.java:
```diff
@@ -0,0 +1,57 @@
+/**
+ * A source module that receives data over WebSocket
+ *
+ * @author Krishnaprasad A S
```
I don't think that this is your official name.
Also you should consider to use it in the Git client config as well.
Right now it comes like `Krishnaprasad <krishna.prasad@flytxt.com>`.

That is just our policy for public community contribution. Thanks for understanding.

krishnaprasadas replied:
"Krishnaprasad A S" is my official name. As you mentioned the git config will be like Krishnaprasad <krishna.prasad@flytxt.com>. So which user name I can use now ?

artembilan commented on spring-cloud-starter-stream-source-websocket/src/main/java/org/springframework/cloud/stream/app/websocket/source/WebsocketSourceProperties.java:
```diff
@@ -0,0 +1,62 @@
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+/**
+ * Command line arguments available for {@link WebsocketSourceConfiguration}
```
That's not command line args. This is "configuration properties": https://docs.spring.io/spring-boot/docs/1.5.9.RELEASE/reference/html/boot-features-external-config.html#boot-features-external-config-typesafe-configuration-properties

artembilan commented on spring-cloud-starter-stream-source-websocket/src/main/java/org/springframework/cloud/stream/app/websocket/source/WebsocketSourceProperties.java:
```diff
@@ -0,0 +1,62 @@
+	private static final String DEFAULT_ALLOWED_ORIGINS = "*";
+
+	/**
+	 * the path on which a WebsocketSource consumer needs to connect. Default is
```
Why does the sentence start not from capital letter ?
  
And somehow the sentence is misleading:
> The path on which server WebSocket handler is exposed

artembilan commented on spring-cloud-starter-stream-source-websocket/src/main/java/org/springframework/cloud/stream/app/websocket/source/WebsocketSourceConfiguration.java:
```diff
@@ -0,0 +1,57 @@
+		return new ServerWebSocketContainer(properties.getPath()).setAllowedOrigins(properties.getAllowedOrigins());
+	}
+
+}
```
New line in the end of each file.
You really can just configure your IDE to do this for you.

artembilan commented on spring-cloud-starter-stream-source-websocket/src/main/java/org/springframework/cloud/stream/app/websocket/source/WebsocketSourceProperties.java:
```diff
@@ -0,0 +1,62 @@
+	 */
+	String allowedOrigins = DEFAULT_ALLOWED_ORIGINS;
+
+	public String getPath() {
```
`@NotEmpty` and `@Validated` on the class level.

artembilan reviewed (changes requested):
Looks like you have missed also `META-INF` stuff: `spring.provides`, `spring-configuration-metadata-whitelist.properties`. See similar in the `spring-cloud-starter-stream-sink-websocket` module. And that really might be a reason of the problem with Docs generation.

Also I would like to see some tests on the matter.

Thanks

Commit:
rebase

krishnaprasadas commented:
@artembilan thanks a lot for your review. Sorry for the delay, I was out of station.
I have some doubts, please see the comments.

artembilan commented:
@krishnaprasadas ,

Thank you for the update anyway, even if the code hasn't been fixed yet!

Well, we really have bumped to the same `NPE` during Docs generation. So, looks like our Maven Doc Plugin isn't compatible with something on the classpath. For example Spring Boot 2.0...

Stay tuned!

/ CC @sobychacko

artembilan commented:
Here we go: https://github.com/spring-cloud/spring-cloud-app-starters-maven-plugins/issues/20 !

artembilan commented:
OK, @krishnaprasadas , we came to the resolution that plugin works well, when all the infrastructure is present and valid. So, as I said before: you need to add `spring-configuration-metadata-whitelist.properties`.

We will consider some improvements in the Doc plugin, but that already not related to your PR...

Commit:
rebase from master & review changes

version change
Readme change
Copyright added
Property for allowedOrigins
Documentation fix

krishnaprasadas commented:
Hi I need a help related to the unit test,

```
@RunWith(SpringRunner.class)
@SpringBootTest(classes = { WebsocketSourceConfiguration.class,
		WebSocketSourceIntegrationTests.class }, webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT, properties = {
				"websocket.path=/some_websocket_path", "websocket.allowedOrigins=*",
				"spring.cloud.stream.default-binder=kafka" })
@EnableAutoConfiguration
public class WebSocketSourceIntegrationTests {

	@Autowired
	private WebsocketSourceProperties properties;

	@LocalServerPort
	private String port;

	@Test
	public void checkCmdlineArgs() {
		assertThat(properties.getPath(), is("/some_websocket_path"));
		assertThat(properties.getAllowedOrigins(), is("*"));
	}

	@Test
	public void testWebSocketStreamSource() throws IOException, InterruptedException {
		StandardWebSocketClient webSocketClient = new StandardWebSocketClient();
		ClientWebSocketContainer clientWebSocketContainer = new ClientWebSocketContainer(webSocketClient,
				"ws://localhost:" + port + "/" + properties.path);
		clientWebSocketContainer.start();
		WebSocketSession session = clientWebSocketContainer.getSession(null);
		session.sendMessage(new TextMessage("foo"));
	}

}
```
I tried the above test, but while building I'm getting the following exception at `WebSocketSession session = clientWebSocketContainer.getSession(null); `

```
testWebSocketStreamSource(org.springframework.cloud.stream.app.websocket.source.WebSocketSourceIntegrationTests)  Time elapsed: 0.498 sec  <<< ERROR!
java.lang.IllegalStateException: javax.websocket.DeploymentException: The HTTP response from the server [200] did not permit the HTTP upgrade to WebSocket
	at org.springframework.integration.websocket.ClientWebSocketContainer.getSession(ClientWebSocketContainer.java:129)
```
As I understand the websocket server is not starting when the test runs.  What is I'm missing?

artembilan commented:
Sounds like that, @krishnaprasadas .

But maybe there is more than just `ws://localhost:" + port + "/" + properties.path` to form the proper URL for the websocket call. You can check in the logs after switching on debug logging level. Or even just run the test in debug mode and place break point in the appropriate Spring Websocket code

artembilan commented:
@krishnaprasadas ,

Any update on the matter?

Maybe you can commit what you have so far and we will fix tests from our side?
That's already too long this PR is jumping here...

Commit:
New dependency for kafka binder and spring boot test

Commit:
Copyright added

krishnaprasadas commented:
@artembilan ,
I have committed test which throws the above mentioned error. If it runs on your side please let me know, such thing happened before also, May be my project setup issue.(Just to understand)

Thanks for your kind support. :)

artembilan commented on spring-cloud-starter-stream-source-websocket/pom.xml:
```diff
@@ -0,0 +1,74 @@
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.cloud</groupId>
+			<artifactId>spring-cloud-starter-stream-kafka</artifactId>
```
Kafka? No, no Kafka as a dependency in this kind of project.
There is a `TestBinder` picked up automatically by Spring Boot.
But what I see you just don't have any auto-configuration detection in your test-case.
I think the `@SpringBootApplication` what we have in the `WebsocketSinkIntegrationTests` is the right way to go.

artembilan commented:
OK, @krishnaprasadas !

Your contribution has been merged to `master` as https://github.com/spring-cloud-stream-app-starters/websocket/commit/c905ef4fb5615e59d9a93ede7c4b50418280c9c6.

And then I have applied this polishing: https://github.com/spring-cloud-stream-app-starters/websocket/commit/72c08da926cd027c2b1038b98259ba9dba5e4aca

See commit message and fixes to the test-case. That really was hard for me to spot that the problem is around security enabled by default, so that might hard to explain it here on the PR how to mitigate.

Anyway I left there some TODO for SockJS, so you may consider more contribution.

Thank you so far!

krishnaprasadas commented:
@artembilan thanks for the merge.

Oops.. security!! yes its very hard, the logs were misleading, at least for me. :)

Yes will definitely try more contributions.

Thank you.
